lucene/core/src/java/org/apache/lucene/search/TaskExecutor.java

```diff
@@ -64,4 +82,26 @@ final <T> List<T> invokeAll(Collection<RunnableFuture<T>> tasks) throws IOExcept
     }
     return results;
   }
+
+  final <C> Task<C> createTask(Callable<C> callable) {
+    return new Task<>(callable);
+  }
+
+  static class Task<V> extends FutureTask<V> {
+    private Task(Callable<V> callable) {
+      super(callable);
+    }
+
+    @Override
+    public void run() {
+      try {
+        Integer counter = runSameThread.get();
+        runSameThread.set(++counter);
```

Comment:
nit: it feels weird to increment when you don't read the value later on
```suggestion
        runSameThread.set(counter + 1);
```